test/validation/api/ipsec/ipsec.c

```diff
@@ -815,10 +858,26 @@ static int ipsec_send_out_one(const ipsec_test_part *part,
 			ev = recv_event(queue, 0);
 			if (ODP_EVENT_INVALID != ev) {
 				CU_ASSERT(odp_event_is_valid(ev) == 1);
+
+				if (ODP_EVENT_IPSEC_STATUS == odp_event_type(ev) &&
```

Comment:
This is needed in the case where the pktio queue is running in scheduled mode and the IPsec destination queue is as well configured for scheduled mode. In this case, the recv_event() call can return an event from either the pktio queue or the IPsec SA destination queue. Hence the handling had to be placed here.

Reply:
This is getting a bit messy. We are seemingly handling an event received from loop pktio, so one would think the event should never be an IPsec status event here. Except that in scheduled mode recv_event() receives from any queue, which is not that obvious.

As a result we have two almost identical branches here with weird checks for scheduled mode and whatnot. I think some refactoring of this code and recv_event() is needed to make the code more readable and maintainable.

One possibility that comes to mind (not sure if it is ok, did not really think it through) could be to have recv_event() always (unless it timeouts) return an event from the queue passed to it, even in scheduled mode. In scheduled mode recv_event() could internally cache received events that came through an unexpected queue. I think it would be sufficient to cache just one event per queue. With this the calling code would be simple and could handle one queue at a time regardless of mode (those "if scheduled mode" checks etc would not be needed).

Reply:
Ack. Added logic to cache at most three scheduled events along with the queue handle via which it was received. With the updated test logic, it runs fine on both linux-gen and cn10k platform.

Reply:
The code changes that make recv_event() take packets only from the specified queue also in scheduled mode could be put in a separate patch that precedes the patch that adds the new test. That would make the changes easier to read.

Reply:
Ack. Made recv_event() change as a separate commit.